### PR TITLE
feat: various metadata improvements

### LINKED ISSUE #2315: Labeling modal
1. Click on X
2. Refresh the page

The labeling modal is visible again. 

image

### LINKED ISSUE #2341: Account name inconsistency in account name / # and label is missing /
**Info:**
 - OS: Nixos
 - Browser: Chrome , version 85.0.4183.83
 - Wallet Version: app/**web**, commit e243afdb73e35fa08f39a7fc08a28b84baa3ac58
 - Firmware Version: model 2.3.3, revision 0d5f00668 
 
**Describe the bug**
In account detail there is **#** missing in account name + there is label missing when labeling is enabled. So account has different name in detail. I know it's a detail but why not to fix it.

**Steps to reproduce:**
1. load accounts and select one
2. and open Account detail and press Show public key
3. Compare account name with Account name in detail

**Expected result:**
You will see account name in account detail

**Actual result:**
You see different account name in account detail

**Screenshots:**
image

image

### LINKED ISSUE #2358: Labeling - Context menu doesnt disappear in receive tab
**Info:**
 - OS: Nixos
 - Browser: Chromium 85.0.4183.83 (Developer Build) (64-bit), Firefox 80.0 (64-bit)
 - Wallet Version: web, commit https://github.com/trezor/trezor-suite/commit/0840609ffa611dc29903e7b5e53b7d73c4381966
 - Firmware Version: model T 2.3.3 , revision 0d5f00668 
 
**Describe the bug**

Labeling context menu doesn't disappear in Receive tab when address has label attached.

**Steps to reproduce:**
1. load wallet and enable labeling
2. go to Receive tab and click on address with label
3. select Confirm on device and observe

**Expected result:**
context menu will disappear after option is selected 
**Actual result:**
context menu will be visible till you click outside that menu

**Screenshots:**
**Chrome**
image

image

**Firefox**
image

### PR DESCRIPTION
**Pull request by mroz22**
Some improvements to metadata. I would like this PR to merge now while it is still small.

- [x] document how metadata works. I believe we should put more effort into documentation. So I started [here](https://github.com/trezor/trezor-suite/blob/feat/metadata3/packages/suite/docs/METADATA.md). I also want to generate screenshots for docs automagically, but this shall have own PR https://github.com/trezor/trezor-suite/pull/2375
- [x] fix https://github.com/trezor/trezor-suite/issues/2357
- [x] resolve #2341
- [x] resolve #2315
- [x] fix #2358 
- [x] do some refactoring

### CODE REVIEW

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (keraf) [inline packages/suite/src/actions/suite/metadataActions.ts]: Console log here should be removed before merge.

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (keraf) [inline packages/suite/src/actions/suite/metadataActions.ts]: I'd be in favor of using a constant (declared in a separate file) here rather than a magic string. You can easily tell the different possible statuses based on that file and it also prevents typos in further work.

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (keraf) [inline packages/suite/src/actions/suite/metadataActions.ts:580]: `device` was destructed on L522, `getState().suite` can be removed.

File: packages/suite/src/views/settings/index.tsx
Comment (keraf) [inline packages/suite/src/views/settings/index.tsx:158]: Same magic string comment as above

File: packages/suite/src/components/suite/Labeling/components/Metadata/index.tsx
Comment (keraf) [inline packages/suite/src/components/suite/Labeling/components/Metadata/index.tsx]: This condition can be simplified to `{!props.payload.value && (`. Also for future reference, starting a condition with a lot of special characters makes it hard to read (in this case 4: `{((!`).

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (keraf) [inline packages/suite/src/actions/suite/metadataActions.ts:572]: Same magic string comment as above.

**Review by keraf — CHANGES_REQUESTED** [review]
Few small changes needed. Double thumbs up for adding docs :+1: :+1:

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (mroz22) [reply packages/suite/src/actions/suite/metadataActions.ts]: that is imho unneeded overhead. typos can not happen as status property is defined as union of string literals "enabled" | "disabled" | "cancelled". VScode also has code completion, so you know immediately possible values without looking to another file. Besides that, this approach (using strings directly) is already used on other places such as opening modals or dispatching notifications.

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (mroz22) [reply packages/suite/src/actions/suite/metadataActions.ts:580]: I don't think you are right on this. I need to get updated value as there is another async action preceding this. One would think reference to object once declared somewhere up would be enough, but it is not. So that code is there on purpose actually.

File: packages/suite/src/components/suite/Labeling/components/Metadata/index.tsx
Comment (mroz22) [reply packages/suite/src/components/suite/Labeling/components/Metadata/index.tsx]: it indeed looks retarded when looking at it after 3 weeks. so retarded that it is making me suspicions why I wrote it like this... hmm...

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (keraf) [reply packages/suite/src/actions/suite/metadataActions.ts]: Indeed, I keep assuming simple strings as I've been mostly working with enums, sorry about that!

File: packages/integration-tests/projects/suite-web/plugins/google.js
Comment (keraf) [inline packages/integration-tests/projects/suite-web/plugins/google.js]: Semi columns missing.

File: packages/suite/src/actions/suite/__fixtures__/metadataActions.ts
Comment (keraf) [inline packages/suite/src/actions/suite/__fixtures__/metadataActions.ts:62]: Is this message going to be displayed to the user? If yes, we should localize it.

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (keraf) [inline packages/suite/src/actions/suite/metadataActions.ts]: Could be simplified:
```
const updatedMetadata = {
    ...account.metadata,
    // always remove metadata  values
    outputLabels: {},
    addressLabels: {},
    accountLabel: '',
};
```

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (keraf) [inline packages/suite/src/actions/suite/metadataActions.ts:170]: We should try to minimize the use of `ts-ignore` unless it's really necessary.

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (keraf) [inline packages/suite/src/actions/suite/metadataActions.ts:168]: Same comment as above. If the string is user facing, we should localize it.

File: packages/suite/src/views/settings/index.tsx
Comment (keraf) [inline packages/suite/src/views/settings/index.tsx]: This should probably be moved into it's own helper function.

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (keraf) [inline packages/suite/src/actions/suite/metadataActions.ts:107]: Very minor: `in` works here but it's inconsistent with how we test for existing properties across the project. Or maybe I'm missing a specific use-case for this operator?

**Review by keraf — CHANGES_REQUESTED** [review]
Looks pretty good, just a few small changes plus the localization and it should be good to merge!

File: packages/suite/src/components/suite/Labeling/components/Metadata/index.tsx
Comment (keraf) [inline packages/suite/src/components/suite/Labeling/components/Metadata/index.tsx]: Nitpicking a bit here but double negatives make conditions harder to read. `labelingImpossible` is used in one place and could be rewritten to `labelingPossible` (with the appropriate condition) and be checked if the value is true rather than false. Not sure if I explained myself well but the gist of it is checking the impossible value is false is harder to read than checking the possible value is true.

File: packages/suite/src/actions/suite/__fixtures__/metadataActions.ts
Comment (mroz22) [reply packages/suite/src/actions/suite/__fixtures__/metadataActions.ts:62]: No, this is test case decription. it(this string goes here, testFnCallback)

File: packages/integration-tests/projects/suite-web/plugins/google.js
Comment (mroz22) [reply packages/integration-tests/projects/suite-web/plugins/google.js]: heh, yes. This file is not covered by eslint at all. Actually are testing setup is quite wild and would require some love. 
For now I have jsust linted this google.js.

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (mroz22) [reply packages/suite/src/actions/suite/metadataActions.ts]: If you wish.. But I don't think it is easier to read. Consider
After your suggestion: 

```javascript
const updatedMetadata = {
  ...account.metadata,
  // always remove metadata  values
  outputLabels: {},
  addressLabels: {},
  accountLabel: '',
};

// and sometimes remove also keys (information we can only if device is connected)
if (keys) {
  updatedMetadata.fileName = '';
  updatedMetadata.aesKey = '';
}

```

My original idea why to write it the way I did was to express 2 stages of deleting values in 2 individual steps. 

```javascript
const updatedMetadata = { ...account.metadata };

// always remove metadata  values
updatedMetadata.outputLabels = {};
updatedMetadata.addressLabels = {};
updatedMetadata.accountLabel = '';

// and sometimes remove also keys (information we can only if device is connected)
if (keys) {
  updatedMetadata.fileName = '';
  updatedMetadata.aesKey = '';
}

```

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (mroz22) [reply packages/suite/src/actions/suite/metadataActions.ts:170]: Yes. I am definitely about to do it. But was not planning to do it in this PR - this ts-ignore was not actually introduced by this PR, you only see it in diff because I reorganized this section of code.
I am planning to rework error handling as part of following PR which will revisit provider logic and authentication and most like will introduce official google lib instead of plain rest api implementation - and this might affect how errors are to be handled.

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (mroz22) [reply packages/suite/src/actions/suite/metadataActions.ts:168]: Yeah, this is probably not going to change and we are going to need some "abstract metadata error" so right, I ll add it.

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (mroz22) [reply packages/suite/src/actions/suite/metadataActions.ts:107]: Using it as narrowing expression for types here. This metadata object may have 2 different shape where one of them dos not have 'key' property. So trying to access it in plain javascript would result in typescript error here.

File: packages/suite/src/views/settings/index.tsx
Comment (mroz22) [reply packages/suite/src/views/settings/index.tsx]: ha, there is a fn for this already. forgot that

File: packages/suite/src/actions/suite/__fixtures__/metadataActions.ts
Comment (keraf) [reply packages/suite/src/actions/suite/__fixtures__/metadataActions.ts:62]: Indeed, I was reviewing going upwards and missed the file title :smile: All good

File: packages/suite/src/components/suite/Labeling/components/Metadata/index.tsx
Comment (mroz22) [reply packages/suite/src/components/suite/Labeling/components/Metadata/index.tsx]: yeah, but it would make definition of labelingPossible harder to read. I would either have to change the conditions to define all cases when it is possible. Generally it is possible when I already have keys (enabled), or do not have them but device is connected which means I can get them. This looks quite OK I think!

```javascript
const labelingPossible = device?.metadata.status === 'enabled' || device?.connected;
```

or negate the expression 
```javascript
const labelingPossible = !(device?.metadata.status !== 'enabled' && !device?.connected);
```
Which I don't like at all :D

Comment (keraf) [reply packages/suite/src/components/suite/Labeling/components/Metadata/index.tsx]: If the condition is too complicated to rewrite (ie: easier to write as a negative), then I guess it makes sense to leave it as is rather than going with the second example (negating the same condition :vomiting_face: ). The first example looks fine if it works for the same cases as the original one.

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (keraf) [reply packages/suite/src/actions/suite/metadataActions.ts]: I see it more as the first step being run in any case, it's clearer to initialize the object like we want it to be rather than reassigning values to its properties right below. Both work, so I'm nitpicking here again :slightly_smiling_face:

File: packages/suite/src/components/suite/Labeling/components/Metadata/index.tsx
Comment (mroz22) [reply packages/suite/src/components/suite/Labeling/components/Metadata/index.tsx]: the first one seems to be the best for me now.

**Review by keraf — APPROVED** [review]
Looks good, waiting for the CI to finish and I'll merge it. Thanks for addressing the comments so quickly :+1:

File: packages/suite/src/actions/suite/metadataActions.ts
Comment (mroz22) [reply packages/suite/src/actions/suite/metadataActions.ts]: I did as you wish but with my heart bleeding :dagger: :heart:

**Pull request merged**
